### ISSUE TITLE
[Feature] Ming-Omni Support Design Spec

### ISSUE THREAD

**Issue by yuan-luo** [sgl-project/sglang-omni#236]
This design spec is to address https://github.com/sgl-project/sglang-omni/issues/191.

## 1. Overview

Integrate the Ming-Omni model (`BailingMM2NativeForConditionalGeneration`) into the sglang-omni framework as a new model pipeline, following the Qwen3-Omni pattern. Ming-Omni is a 100B-total / 6B-active MoE multimodal model supporting text, image, video, and audio understanding, plus text, speech, and image generation.

**Initial scope**: Text and audio input → text and speech output, with basic TTS (default voice, no voice cloning, no streaming audio).

---

## 2. Architecture Decision

Ming-Omni's Thinker and Talker are **loosely coupled via text** — the Thinker generates text tokens, which are decoded to a text string and passed to the Talker for independent TTS. This differs from Qwen3-Omni where the Talker consumes hidden states from the Thinker.

This makes the integration simpler: no hidden state relay is needed between Thinker and Talker stages.

---

## 3. Pipeline Architecture (6 stages)

```
Preprocessing (CPU) → Audio Encoder (GPU) → Aggregate (CPU)
→ Thinker/SGLang (GPU) → [fan-out] → Decode (CPU, terminal)
                                     → Talker (GPU, terminal)
```

```
  ┌───────────────┐    ┌───────────────┐    ┌─────────────┐
  │ Preprocessing │───▶│ Audio Encoder │───▶│  Aggregate  │
  │   (CPU)       │    │ (Whisper+Proj)│    │ (merge all) │
  └───────────────┘    │   (GPU)       │    └──────┬──────┘
                       └───────────────┘           │
                                                   ▼
                                            ┌──────────────┐
                                            │   Thinker    │
                                            │(BailingMoeV2)│
                                            │  (SGLang)    │
                                            └───┬──────┬───┘
                                                │      │
                                                │      │ text output
                                                ▼      ▼
                                       ┌─────────┐  ┌──────────┐
                                       │ Decode  │  │  Talker  │
                                       │ (text)  │  │(own LLM +│
                                       │terminal │  │CFM+DiT+  │
                                       └─────────┘  │AudioVAE) │
                                                    │terminal  │
                                                    └──────────┘
```

### Stage Details

| # | Stage | Device | Executor Type | Input | Output |
|---|-------|--------|--------------|-------|--------|
| 1 | `preprocessing` | CPU | PreprocessingExecutor | ChatCompletionRequest | MingPipelineState (input_ids, mel features) |
| 2 | `audio_encoder` | GPU | DirectModelExecutor | Mel-spectrogram features | Audio embeddings (4096-dim) |
| 3 | `aggregate` | CPU | PreprocessingExecutor | Preprocessing + encoder outputs | Merged thinker input |
| 4 | `thinker` | GPU:0 | EngineExecutor (SGLang) | Merged embeddings | Generated text token IDs |
| 5 | `decode` | CPU | PreprocessingExecutor (terminal) | Token IDs | Text events (OmniEvent) |
| 6 | `talker` | GPU:1 | DirectModelExecutor (terminal) | Decoded text string | Audio waveform |

### Routing

- `preprocessing` → `audio_encoder` (if audio present), else → `aggregate`
- `audio_encoder` → `aggregate`
- `aggregate` → `thinker`
- `thinker` → fan-out: [`decode`, `talker`]
- `decode` → terminal
- `talker` → terminal

### GPU Placement (default)

```python
gpu_placement = {
    "thinker": 0,      # BailingMoeV2 (main LLM)
    "audio_encoder": 0, # Whisper encoder (shared with thinker GPU)
    "talker": 1,        # BailingTalker2 (own LLM + CFM + VAE)
}
```

---

## 4. File Structure

```
sglang_omni/models/ming_omni/
├── __init__.py                    # Package exports
├── config.py                      # MingOmniPipelineConfig, MingOmniSpeechPipelineConfig
│                                  # EntryClass = MingOmniPipelineConfig
├── hf_config.py                   # Pydantic models for BailingMM2Config, BailingMoeV2Config,
│                                  # WhisperEncoderConfig, BailingTalkerConfig
├── io.py                          # MingPipelineState dataclass, MingOmniEvent
├── pipeline/
│   ├── __init__.py
│   ├── stages.py                  # create_preprocessing_executor()
│   │                              # create_audio_encoder_executor()
│   │                              # create_aggregate_executor()
│   │                              # create_sglang_thinker_executor()
│   │                              # create_talker_executor()
│   │                              # create_decode_executor()
│   ├── next_stage.py              # Stage name constants + routing functions
│   └── merge.py                   # merge_for_thinker(), decode_events()
├── components/
│   ├── __init__.py
│   ├── preprocessor.py            # MingPreprocessor (BailingTokenizer + audio feat extraction)
│   ├── audio_encoder.py           # MingAudioEncoder (WhisperEncoder + Conv1d + MLP projection)
│   ├── thinker.py                 # BailingMoeV2ForCausalLM SGLang model class
│   ├── talker_executor.py         # MingTalkerExecutor (wraps BailingTalker2)
│   └── audio_vae.py               # AudioVAE decoder (from talker weights)
```

---

### 5.1 Preprocessing

**Responsibilities:**
- Apply BailingMM2's chat template (`<role>HUMAN</role>...`, `<role>ASSISTANT</role>...`)
- Tokenize text with BailingTokenizer (vocab 157,184)
- For audio inputs: load audio → resample → compute mel-spectrogram (128 bins)
- Insert `<audio>`, `</audio>`, `<audioPatch>` placeholder tokens
- Output: `MingPipelineState` with `input_ids`, `audio_features`, `audio_feature_lengths`, `audio_placeholder_positions`

**Reference**: `processing_bailingmm2.py` (592 lines), `audio_processing_bailingmm2.py` (378 lines)

### 5.2 Audio Encoder

**Architecture** (from Ming's `config.json`):
- WhisperAudioEncoder: 128 mel bins, 32 layers, 1280 hidden dim, 20 heads, max 15000 frames
- Conv1d projection: 1280 → 4096, kernel=3, stride=2
- 2-layer MLP with GELU: 4096 → 4096
- Output normalization: `F.normalize(dim=2)`

**Weight extraction**: Load from `BailingMM2` checkpoint, extract `audio.*` and `linear_proj_audio.*` parameters.

**Reference**: `modeling_whisper_encoder.py`, `modeling_bailingmm2.py:62-88`

### 5.3 Aggregate

**Responsibilities:**
- Receive outputs from preprocessing and audio encoder
- Patch audio embeddings into `<audioPatch>` token positions in the embedding sequence
- Build the final thinker input: `input_ids` with audio embeddings replacing placeholder tokens
- Handle variable-length audio with padding/masking

**Reference**: `modeling_bailingmm2.py:extract_audio_feature()`, `modeling_utils.py:patch_continuous_features()`

### 5.4 Thinker (SGLang BailingMoeV2)

**New SGLang model class**: `BailingMoeV2ForCausalLM`

**Architecture** (from `config.json`):

| Parameter | Value |
|-----------|-------|
| num_hidden_layers | 32 |
| num_attention_heads | 32 |
| num_key_value_heads | 4 |
| hidden_size | 4096 |
| intermediate_size | 9216 |
| moe_intermediate_size | 1024 |
| num_experts | 256 |
| num_experts_per_tok | 8 |
| num_shared_experts | 1 |
| router_type | MultiRouter |
| n_group | 8 |
| topk_group | 4 |
| moe_router_topk_scaling_factor | 2.5 |
| use_expert_bias | true |
| partial_rotary_factor | 0.5 |
| rope_theta | 2400000 |
| rope_scaling.type | video_rope |
| use_qk_norm | true |
| hidden_act | silu |
| rms_norm_eps | 1e-6 |
| vocab_size | 157184 |
| first_k_dense_replace | 1 |

**Implementation notes:**
- First layer is dense (first_k_dense_replace=1), rest are MoE
- `MultiRouter` with grouped routing: divide 256 experts into `n_group=8` groups of 32, select `topk_group=4` groups by highest average router score, then select top-k within each selected group to reach `num_experts_per_tok=8` total experts. Apply `moe_router_topk_scaling_factor=2.5` scaling to router logits before softmax
- `partial_rotary_factor=0.5`: Apply RoPE to first half of head dim only
- `video_rope` scaling: Custom RoPE for video temporal positions (needs investigation)
- Uses SGLang's `RadixAttention` for paged KV cache
- Uses SGLang's `FusedMoE` kernel for efficient expert computation

**Split loading**: Load LLM body and LM head separately (following Qwen3-Omni thinker pattern) for memory efficiency.

### 5.5 Decode

**Responsibilities:**
- Convert generated token IDs to text
- Emit `OmniEvent` objects: `text_delta`, `text_final`
- Handle special tokens and EOS

### 5.6 Talker (BailingTalker2)

**Architecture**: Self-contained TTS module with its own LLM + CFM + DiT + AudioVAE.

**Internal components:**
- **Talker LLM**: Qwen2 (dense) or BailingMoe (MoE), loaded from `{model_path}/talker/llm`
- **CFM** (Conditional Flow Matching): DiT-based diffusion with configurable steps
- **Aggregator**: Maps audio latents → LLM embedding space
- **Stop head**: Binary classifier on hidden state → stop signal
- **AudioVAE**: Latent → waveform decoder, loaded from `{model_path}/talker/vae`
- **Text normalizer** (TalkerTN): Chinese + English text normalization

**Executor design**: `MingTalkerExecutor` wraps BailingTalker2 as a standalone executor:
- Receives decoded text string from thinker output
- Runs `omni_audio_generation(tts_text=text, voice_name="DB30")`
- Collects audio latent patches → AudioVAE → waveform
- Returns waveform as payload data

**CUDA graph support**: The talker uses CUDA graphs for both:
- LLM decode steps (StaticCache-based)
- CFM sampling (CFMGraphExecutorPool)

**Concurrency**: Currently limited to `max_conc` concurrent requests via `CFMGraphExecutorPool`.

---

### Pipeline Config

```python
class MingOmniSpeechPipelineConfig(PipelineConfig):
    architecture: ClassVar[str] = "BailingMM2NativeForConditionalGeneration"
    model_path: str                              # Path to Ming-Omni model weights
    talker_model_path: str = ""                  # Defaults to {model_path}/talker
    entry_stage: str = "preprocessing"
    terminal_stages: list[str] = ["decode", "talker"]
    gpu_placement: dict[str, int] = {
        "thinker": 0,
        "audio_encoder": 0,
        "talker": 1,
    }
    relay_backend: str = "shm"
```

### Text-only Config (no talker)

```python
class MingOmniPipelineConfig(PipelineConfig):
    architecture: ClassVar[str] = "BailingMM2NativeForConditionalGeneration"
    model_path: str
    entry_stage: str = "preprocessing"
    terminal_stages: list[str] = ["decode"]
    stages: list[StageConfig] = [
        # preprocessing, audio_encoder, aggregate, thinker, decode
    ]
```

---

## 7. API Compatibility

Reuse existing sglang-omni OpenAI-compatible API:
- `POST /v1/chat/completions` — text + audio input, text + audio output
- `POST /v1/audio/speech` — TTS endpoint (routes directly to talker)

Request format:
```json
{
    "model": "ming-omni",
    "messages": [
        {"role": "user", "content": [
            {"type": "text", "text": "Describe this audio"},
            {"type": "audio_url", "audio_url": {"url": "..."}}
        ]}
    ],
    "audio": {"voice": "default", "format": "wav"}
}
```

---

## 8. Weight Loading Strategy

Ming-Omni's weights are organized as a single HuggingFace checkpoint:

```
model_path/
├── config.json                    # BailingMM2Config
├── tokenizer.json                 # BailingTokenizer
├── tokenizer_config.json
├── model-*.safetensors            # All weights (LLM + vision + audio)
├── talker/
│   ├── config.json                # BailingTalkerConfig
│   ├── model weights
│   └── vae/
│       ├── config.json
│       └── model weights
└── campplus.onnx                  # Speaker embedding (not used initially)
```

**Loading strategy:**
1. **Audio encoder**: Extract `audio.*` + `linear_proj_audio.*` from main checkpoint
2. **Thinker**: Extract `model.*` (BailingMoeV2) from main checkpoint, load into SGLang model class
3. **Talker**: Load from `{model_path}/talker/` as self-contained BailingTalker2
4. **AudioVAE**: Load from `{model_path}/talker/vae/` separately

---

## 9. Key Risks & Mitigations

| Risk | Mitigation |
|------|-----------|
| BailingMoeV2's `MultiRouter` not supported by SGLang's FusedMoE | Implement custom routing layer; may need to adapt SGLang's grouped top-k |
| `partial_rotary_factor=0.5` not standard in SGLang | Implement partial RoPE in the attention layer |
| `video_rope` scaling type unknown | Investigate Ming's source; may only matter for video (out of scope) |
| Talker CUDA graph pool limits concurrency | Accept initial limitation; optimize in follow-up |
| Large model (100B total params) memory pressure | Use `gpu_placement` to distribute across GPUs; thinker loads only active params via MoE |

---

## 10. Out of Scope (Future Work)

- Image/video input support
- Image generation output
- Voice cloning (CampPlus speaker embeddings)
- Controllable TTS attributes (emotion, speed, pitch, dialect, BGM)
- Streaming audio output (SSE)
- Multi-GPU tensor parallelism for thinker
- Talker concurrency optimization
